src/wh_server_keystore.c

```diff
@@ -0,0 +1,293 @@
+
+int hsmGetUniqueId(whServerContext* server)
+{
+    int i;
+    int ret = 0;
+    uint16_t id;
+    /* try every index, unless both cache and nvm are full one will work */
+    for (id = 1; id < WOLFHSM_NUM_RAMKEYS + 1; id++) {
+        /* check against chache keys */
+        for (i = 0; i < WOLFHSM_NUM_RAMKEYS; i++) {
+            if (id == server->cache[i].meta->id)
+                break;
+        }
+        /* try again if match */
+        if (i < WOLFHSM_NUM_RAMKEYS)
+            continue;
+#if 0
+        /* check against nvm keys */
+        for (i = 0; i < WOLFHSM_KEYSLOT_COUNT && ret == 0; i++) {
+            if (ctx->nvmMetaCache[i].id == id) {
+                break;
+            }
+        }
+        /* try again if match */
+        if (i < WOLFHSM_KEYSLOT_COUNT)
+            continue;
+#endif
+        /* break if our id didn't match either cache or nvm */
+        break;
+    }
+    /* if we've run out of space, return MEMORY_E */
+    if (ret == 0 && id >= WOLFHSM_NUM_RAMKEYS + 1)
+        ret = MEMORY_E;
+    /* ultimately, return found id */
+    if (ret == 0)
+        ret = id;
+    return ret;
+}
+
+int hsmCacheKey(whServerContext* server, whNvmMetadata* meta, uint8_t* in)
+{
+    int i;
+    int foundIndex = -1;
+    /* make sure id is valid */
+    if (meta->id == 0)
+        return BAD_FUNC_ARG;
+    /* make sure the key fits in a slot */
+    if (meta->len > WOLFHSM_NVM_MAX_OBJECT_SIZE)
+        return BUFFER_E;
+    for (i = 0; i < WOLFHSM_NUM_RAMKEYS; i++) {
+        /* check for empty slot or rewrite slot */
+        if (foundIndex == -1 && (server->cache[i].meta->id == 0 ||
+            server->cache[i].meta->id == meta->id)) {
+            foundIndex = i;
+        }
+    }
+#if 0
+    /* if no empty slots, check for a commited key we can evict */
+    if (foundIndex == -1) {
+        for (i = 0; i < WOLFHSM_CACHE_COUNT; i++) {
+            if (ctx->cache[i].commited == 1) {
+                foundIndex = i;
+                break;
+            }
+        }
+    }
+    /* return error if we are out of cache slots */
+    if (foundIndex == -1)
+        return MEMORY_E;
+    ctx->cache[foundIndex].commited = 0;
+    /* check if the key is in nvm, mark as commited */
+    for (i = 0; i < WOLFHSM_KEYSLOT_COUNT; i++) {
+        /* check for duplicate id, shouldn't cache a key in nvm */
+        if (ctx->nvmMetaCache[i].id == meta->id)
+            ctx->cache[foundIndex].commited = 1;
+    }
+#endif
+    /* write key if slot found */
+    XMEMCPY((uint8_t*)server->cache[foundIndex].buffer, in, meta->len);
+    XMEMCPY((uint8_t*)server->cache[foundIndex].meta, (uint8_t*)meta,
+        sizeof(whNvmMetadata));
+#ifdef WOLFHSM_SHE_EXTENSION
+    /* if this was RAM_KEY, set the global so we can export */
+    if (meta->id == WOLFHSM_SHE_TRANSLATE_KEY_ID(WOLFHSM_SHE_RAM_KEY_ID))
+        hsmSheRamKeyPlain = 1;
+#endif
+    return 0;
+}
+
+int hsmReadKey(whServerContext* server, whNvmMetadata* meta, uint8_t* out)
```

Comment:
I found it a bit confusing that this function secretly uses the len field of metadata as an in/out argument to set the max length. Could we document this as a side-effect in a comment, or provide separate max length arguments? I was staring at the calling context trying to figure out how it worked for a while before realizing that the caller must initialize `meta` with a max output length before invoking this function, but then it overwrites the entire `meta` struct with the cached version once found. I see how it works now, but I find that clever tricks like this just make code harder to grok on the first pass, vs. using an explict max len argument that makes it immediately obvious what is happening to the reader (alternatively, just document the behavior in a comment).